openssl/src/pkcs7.rs

```diff
@@ -279,17 +392,191 @@ impl Pkcs7Ref {
+    }
+
+    /// Add the signer certificate to a PKCS#7 structure
+    ///
+    /// `cert` is the signer's certificate.
+    ///
+    /// This method moves the ownership of `cert` to the PKCS7 structure.
+    ///
+    #[corresponds(PKCS7_add_certificate)]
+    ///
+    pub fn add_certificate(&self, cert: X509) -> Result<(), ErrorStack> {
+        unsafe {
+            cvt(ffi::PKCS7_add_certificate(self.as_ptr(), cert.as_ptr()))?;
+            mem::forget(cert);
+            Ok(())
+        }
+    }
+
+    /// Add signature information to the PKCS#7 structure.
+    ///
+    /// `cert` is the signer's certificate. `pkey` is the signer's (private) key. `algorithm` is
+    /// the hash algorithm to be used.
+    ///
+    /// Returns a signer info structure, which can be used to add signed attributes. `cert` is not
+    /// consumed by this method (actually by `PKCS7_add_signature`), but `pkey` is. Create a clone
+    /// before calling `add_signature()`, if you need the key later.:
+    /// ```
+    /// use openssl::hash::MessageDigest;
+    /// use openssl::pkcs7::Pkcs7;
+    /// use openssl::pkey::PKey;
+    /// use openssl::rsa::Rsa;
+    /// use openssl::x509::X509;
+    ///
+    /// let cert = X509::from_pem("-----BEGIN CERTIFICATE-----\n\
+    /// MIICsTCCAZmgAwIBAgIBADANBgkqhkiG9w0BAQsFADAcMRowGAYDVQQDDBFJU0VD\n\
+    /// IFRlc3QgUm9vdCBDQTAeFw0yMTEyMTUwOTM4MDVaFw0zMTEyMTMwOTM4MDVaMBwx\n\
+    /// GjAYBgNVBAMMEUlTRUMgVGVzdCBSb290IENBMIIBIjANBgkqhkiG9w0BAQEFAAOC\n\
+    /// AQ8AMIIBCgKCAQEAt/0hnw55JtLXN4QjFhmkZ5mSSWQM3Zlz8+n9wq99jkDXEUJ8\n\
+    /// YGoHxpBTb9vz4BXwVcR/3GWtdJ/h5VQinBYcZLwtz5iel4IC7pl40a8Kcco3hm6U\n\
+    /// +qve9wz2YS8coQ1+zQ/pqKxDOLN60BYVkuxZeC4yrg8ovL5YftKzeVmFUjDJ/vdI\n\
+    /// RyDScFpso2UoW6mklW/C94ciJH6O9m9dd2nWow1vUfHJlAEm8nPRxJRQH1bROxmc\n\
+    /// 5hLHMjY5wFD/jo31jzxSeZxNhLTKDwn8nG1AGoHaudUNtAF25tzcU6nLKt6BIS+g\n\
+    /// wqavL3kwp+1O/GwLPQb4xXmPs2+f2M08XMbVKwIDAQABMA0GCSqGSIb3DQEBCwUA\n\
+    /// A4IBAQCKWDHBLlLPDmb1C5FOcJ/wqdwCzkbZEBs3qsiim5EDkt9+nqAfwyn0K+G8\n\
+    /// BLJU+6kgeGW1Z0t3RqJdAq2r+7bfMVF8ubJ4zEu5xJAz9UppI4XZ8sZ8iS5rZ3hq\n\
+    /// zjU+7G8Lnu9gEh18q5foi59Wx0jjMyIOWh8O9j3P0JjLxAR8v4rKlYp89/A+vWfb\n\
+    /// TAj31LhWWTa0kiDP9Wd8cMWCjurv7Wq7U4K9gHMHpmUclxs1ByHtFdd61OXSdfBx\n\
+    /// P1Te4tjxQpVo5zURkwOfZoOC6ikAOYoTAYNHP/qwsX2+KnL/JeGCcU4WtoWIp8mt\n\
+    /// HkHgG5EBGoJmLFNFeQWB2yaqAhfD\n\
+    /// -----END CERTIFICATE-----".as_bytes()).unwrap();
+    /// let rsa = Rsa::generate(2048).unwrap();
+    /// let signer_key = PKey::from_rsa(rsa).unwrap();
+    /// let signer_key_clone = signer_key.clone();
+    /// let pkcs7 = Pkcs7::new().unwrap();
+    /// let signer_info = &pkcs7.add_signature(
+    ///     &cert,
+    ///     signer_key_clone,
+    ///     MessageDigest::sha256()
+    /// );
+    ///  ```
+    ///
+    #[corresponds(PKCS7_add_signature)]
+    pub fn add_signature<PT>(
+        &self,
+        cert: &X509Ref,
+        pkey: PKey<PT>,
```

Comment:
This does not take ownership of the pkey.

Reply:
But [OpenSSL increments reference count](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/pkcs7/pk7_lib.c#L324) on `pkey`. How do we handle that? From OpenSSLs point of view, we have 2 owners after `ffi::PKCS7_add_signature` was called: the `PKey`, that was passed and the `Pkcs7`.

Reply:
This method should take a `&PKeyRef<PT>`.

Reply:
Got it. Fixed.